x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/TopValuesListLongAggregator.java

```diff
@@ -0,0 +1,222 @@
+
+        private GroupingState(int limit, boolean ascending) {
+            this.queuesByGroup = new HashMap<>();
+            this.queueComparator = ascending ? Comparator.reverseOrder() : Comparator.naturalOrder();
+            this.finalSortComparator = ascending ? Comparator.naturalOrder() : Comparator.reverseOrder();
+            this.limit = limit;
+        }
+
+        public void add(int groupId, long value) {
+            var queue = queuesByGroup.computeIfAbsent(groupId, k -> new Queue(queueComparator, limit));
+
+            queue.safeAdd(value, limit);
+        }
+
+        public void merge(int groupId, GroupingState other, int otherGroupId) {
+            var queue = queuesByGroup.computeIfAbsent(groupId, k -> new Queue(queueComparator, limit));
+            var otherQueue = other.queuesByGroup.get(otherGroupId);
+
+            if (otherQueue == null) {
+                return;
+            }
+
+            for (var value : otherQueue) {
+                queue.safeAdd(value, limit);
+            }
+        }
+
+        void toIntermediate(Block[] blocks, int offset, IntVector selected, DriverContext driverContext) {
+            blocks[offset] = toBlock(driverContext.blockFactory(), selected);
+        }
+
+        Block toBlock(BlockFactory blockFactory, IntVector selected) {
+            if (queuesByGroup.isEmpty()) {
+                return blockFactory.newConstantNullBlock(selected.getPositionCount());
+            }
+            try (LongBlock.Builder builder = blockFactory.newLongBlockBuilder(selected.getPositionCount())) {
+                for (int s = 0; s < selected.getPositionCount(); s++) {
+                    int selectedGroup = selected.getInt(s);
+
+                    var queue = queuesByGroup.get(selectedGroup);
+
+                    if (queue == null) {
+                        builder.appendNull();
+                        continue;
+                    }
+
+                    if (queue.size() == 1) {
+                        builder.appendLong(queue.top());
+                        continue;
+                    }
+
+                    builder.beginPositionEntry();
+
+                    StreamSupport.stream(queue.spliterator(), false)
+                        .sorted(finalSortComparator)
+                        .forEach(builder::appendLong);
+
+                    builder.endPositionEntry();
+                }
+                return builder.build();
+            }
+        }
+
+        void enableGroupIdTracking(SeenGroupIds seen) {
+            // we figure out seen values from nulls on the values block
+        }
+
+        void close() {
+            // Nothing to do
+        }
+    }
+
+    private static class Queue extends PriorityQueue<Long> {
+        private final Comparator<Long> comparator;
+
+        public Queue(Comparator<Long> comparator, int limit) {
+            super(limit + 1);
+            this.comparator = comparator;
+        }
+
+        public void safeAdd(long value, int limit) {
```

Comment:
In general I'd just use `insertWithOverflow` with Lucene's priority queue. That's not really good if the thing you are offering has to get boxed though. Which is why I wanted `BucketedSort`.